---
title: "Lab_8_EvoGeno_BenEarle"
author: "Ben Earle"
date: "10/25/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Exercises 1

### In exercise 1 we learn use NGPhylogeny.FR, which is a convenient online source that provides a workflow for phylogenetic analysis. These analyses are useful for multiple sequence alignment, phylogenetic analysis, and tree visualization. In order to create a phylogenetic tree, we must first assemble a group of sequences to evaluate. Then, the workflow aligns the sequences to that distance can be calculated between the sequences so that we can then understand the relationship between the sequences. Once this is accomplished, we can then visualize the relationships in a phylogenetic tree like the example provided below. 

[Phylogenetic Tree](Screen Shot 2020-10-25 at 1.02.08 PM.png)

# Exercise 2

### We now take a deeper dive into phylogenetic analysis by learning to use CIPRES. CIPRES is a public resource for inference of large phylogenetic trees. 

[Tol output](TOL_output.mafft)

[Fasttree result](fastree_result.tre)


# Exercise 3 - Visualizing & analyzing phylogenetic trees w/ R + ggtree

```{r, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggtree")
```

```{r, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggimage")
```

```{r, include=FALSE}
library(tidyverse)
library(ggtree)
# added treeio and ggimage library
library(treeio)
library(ggimage)
# need to install libmagick 
#sudo apt-get install libmagick++-dev
```

## Basic Trees

```{r}
tree <- read.tree("tree_newick.nwk")
tree
```

```{r}
ggtree(tree)
```

### Add a scale
```{r}
ggtree(tree) + theme_tree2()
```

### The default is to plot a phylogram, where the x-axis shows the genetic change / evolutionary distance. If you want to disable scaling and produce a cladogram instead, set the branch.length="none" option inside the ggtree()

```{r}
ggtree(tree, branch.length="none")
```

### cladogram (no branch scaling) using thick blue dotted lines

```{r}
ggtree(tree, branch.length="none", color="blue", size=2, linetype=3)
```

# Exercise 1 
## Part 1 - slanted phylogenetic tree
```{r}
ggtree(tree, branch.length="none", color="blue", size=2, linetype=3, layout = "slanted")
```
## Part 2 - circular phylogenetic tree
```{r}
ggtree(tree, branch.length="none", color="blue", size=2, linetype=3, layout = "circular")
```
## Part 3 - circular unscaled claudogram with thick red line

```{r}
ggtree(tree, branch.length="none", color="red", size=2, linetype=3, layout = "circular")
```
# Other Tree Geoms
###In order to add additional layers, we must first create a plot object, p, to store the basic layout of a ggplot. We will then add node and tip points, then label the tips (all in the same code chunk).
```{r}
p <- ggtree(tree) 
```

```{r}
p + geom_nodepoint() +
  geom_tippoint() +
  geom_tiplab()
```
## Exercise 2
```{r}
p + geom_nodepoint(color = "yellow", alpha = NA) +
  geom_tippoint(pch = 18, color = "purple") +
  geom_tiplab() +
  ggtitle("Exercise 2 Figure")
```
## Internal node number
```{r}
ggtree(tree) + geom_text(aes(label=node), hjust=-.3)
```

```{r}
ggtree(tree) + geom_tiplab()
```

```{r}
MRCA(tree, c("C", "E"))
```

```{r}
MRCA(tree, c("G", "H"))
```

```{r}
# MRCA(tree, tip=c("C", "E"))
# MRCA(tree, tip=c("G", "H"))
```


## Labeling Clades
### We can use geom_cladelabel() to add another geom layer to annotate a selected clade with a bar indicating the clade with a corresponding label. You select the clades using the internal node number for the node that connects all the taxa in that clade.

### Annotation of the Clade with the most recent common ancestor between taxa C and E (internal node 17):
```{r}
ggtree(tree) + 
  geom_cladelabel(node=17, label="Some random clade", color="red")
```
### Adding back the tips
```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8)
```
### adding another label for the clade connecting taxa G and H (internal node 21).
```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8, align=TRUE) + 
  geom_cladelabel(node=21, label="A different clade", 
                  color="blue", offset=.8, align=TRUE)
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8, align=TRUE) + 
  geom_cladelabel(node=21, label="A different clade", 
                  color="blue", offset=.8, align=TRUE) + 
  theme_tree2() + 
  xlim(0, 70) + 
  theme_tree()
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_hilight(node=17, fill="gold") + 
  geom_hilight(node=21, fill="purple")
```

## Connecting Taxa
```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_taxalink("E", "H", color="blue3") +
  geom_taxalink("C", "G", color="orange2", curvature=-.9)
```

## Exercise 3 
```{r}
MRCA(tree, c("B", "C"))
```

```{r}
MRCA(tree, c("L", "J"))
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Superclade label 17", 
                  color="red2", offset=.8, align=TRUE) + 
  geom_highlight(node=19, fill = "purple") +
  geom_highlight(node=23, fill = "Gold") +
  geom_taxalink("C", "E", linetype=2, color="dark grey") +
  geom_taxalink("G", "J", linetype=2, color="dark grey") +
  theme_tree() +
  xlim(0, 70)
```

# Advanced Tree Annotation

```{r}
tree <- read.beast("flu_tree_beast.tree")
```

```{r}
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() 
```

```{r}
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() + 
  geom_tiplab(align=TRUE, linesize=.5) + 
  xlim(1990, 2020)
```

```{r}
msaplot(p=ggtree(tree), fasta="flu_aasequence.fasta", window=c(150, 175))
```

# Bonus!

## Many Trees
```{r}
set.seed(42)
trees <- lapply(rep(c(10, 25, 50, 100), 3), rtree)
class(trees) <- "multiPhylo"
ggtree(trees) + facet_wrap(~.id, scale="free", ncol=4) + ggtitle("Many trees. Such phylogenetics. Wow.")
```

## Plot trees with other date
```{r}
tree <- rtree(30)
```
### Make original plot
```{r}
p <- ggtree(tree)
```
### generate some random values for each tip label in the data
```{r}
d1 <- data.frame(id=tree$tip.label, val=rnorm(30, sd=3))
```
### Make a second plot with the original, naming the new plot "dot", 
### using the data you just created, with a point geom.
```{r}
p2 <- facet_plot(p, panel="dot", data=d1, geom=geom_point, aes(x=val), color='red3')
```
### Make some more data with another random value.
```{r}
d2 <- data.frame(id=tree$tip.label, value = abs(rnorm(30, mean=100, sd=50)))
```
### Now add to that second plot, this time using the new d2 data above, 
### This time showing a bar segment, size 3, colored blue.
```{r}
p3 <- facet_plot(p2, panel='bar', data=d2, geom=geom_segment, 
           aes(x=0, xend=value, y=y, yend=y), size=3, color='blue4') 
```
### Show all three plots with a scale
```{r}
p3 + theme_tree2()
```


# Exercise 4 original exercises
```{r}
tree <- read.tree("fastree_result.tre")
tree
```
```{r}
ggtree(tree) +  
  geom_nodelab() +
  geom_tiplab(hjust=-.1, size = 3) +
  geom_hilight(node=34, fill="gold", extend = 0.5) 
```


## References

title: "Evolutionary Genomics & Bioinformatics"
author: "Jeffrey Blanchard"
date: "2/26/2020"
output: html_document
bibliography: data/references.bib

title: "Visualizing and Annotating Phylogenetic Trees with R+ggtree"
Author: "Stephen Turner"
output: html_document
bibliography: https://4va.github.io/biodatasci/r-ggtree.html

title: "ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data"
author: "Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam"
date: "2017"
output: html_document
bibliography: https://doi.org/10.1093/molbev/msy194

title: "Expansion of genotypic diversity and establishment of 2009 H1N1 pandemic-origin internal genes in pigs in China."
author: "Huey Liang"
date: "September 2014"
output: html_document
bibliography:https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4178866/


  
Application written in R [@RCoreTeam] using the Shiny framework [@Chang2015].
